### ISSUE TITLE
[model.load_adapter_fusion] Cannot load pre-trained adapter fusion into model

### ISSUE THREAD

**Issue by Yujin-Yujin** [Adapter-Hub/adapter-transformers#74]
Even all the layer is loaded to the model, however, the weights are not applied at all.
is there any version issue or did I miss something?

    

    model.load_adapter("/home/test/siqa_default", "text_task",config=PfeifferConfig(), with_head=False)
    model.load_adapter("/home/test/a", "text_task",config=PfeifferConfig(), with_head=False)
    model.load_adapter("/home/test/b", "text_task",config=PfeifferConfig(), with_head=False)
    model.load_adapter("/home/test/c", "text_task",config=PfeifferConfig(), with_head=False)
    model.load_adapter("/home/test/d", "text_task",config=PfeifferConfig(), with_head=False)

    adapter_names = [
            [
                "siqa_default",
                "a",
                "b",
                "c",
                "d"
            ]
        ]

    # pre-trained fusion_path 
        fusion_path = ="/home/test/fusion/siqa_defaut,a,b,c,d"
        model.load_adapter_fusion(fusion_path)

    # test_dataset
    test_dataset = (
        MultipleChoiceDataset(
            data_dir=data_args.data_dir,
            tokenizer=tokenizer,
            task=task_type,
            max_seq_length=data_args.max_seq_length,
            overwrite_cache=data_args.overwrite_cache,
            mode=Split.test,
        )
    )

    def compute_metrics(p: EvalPrediction) -> Dict:
        preds = np.argmax(p.predictions, axis=1)
        return {"acc": simple_accuracy(preds, p.label_ids)}
 
    # Initialize our Trainer
    trainer = Trainer(
        model=model,
        args=training_args,
        compute_metrics=compute_metrics,
        adapter_names=adapter_names
    )

**Comment by calpt** [comment]
This might be an issue which was fixed recently. Please make sure you have the latest version PyPI version of adapter-transformers (v1.0.1) or the latest version from the master branch of this repo installed.

If the issue still persists: Do you get an error message? If yes, please add the full message here.

**Comment by Yujin-Yujin** [comment]
I've already tried new version adapter-transformers 1.0.1, and I got the issue that I mentioned above.
I did not get any error message, but I couldn't get performance at all.
For example, when I trained fusion and tested at last, the test score was 61. 
However, when I load the saved fusion and tested, I got the score 20, which is lower than not using adapter.

I think load_adapter_fusion method does not work at all, or my saved fusion is wrong. 
I trained fusion with the code below.
    from transformers.adapter_config import PfeifferConfig

    # config = AdapterConfig.load("pfeiffer")
    model.load_adapter("multinli@ukp", "text_task", config=PfeifferConfig())

    model.load_adapter("/home/yujin/adapters/default/siqa_default", config=PfeifferConfig(), with_head=False)

    adapter_names = [
        [
            "mnli",
            "siqa_default"
        ]
    ]

    model.add_fusion(adapter_names[0], "dynamic")
    model.train_fusion(adapter_names)
    ...
        # Initialize our Trainer
    trainer = Trainer(
        model=model,
        args=training_args,
        train_dataset=train_dataset,
        eval_dataset=eval_dataset,
        compute_metrics=compute_metrics,
        do_save_full_model= False,
        do_save_adapter_fusion= True,
        adapter_names=adapter_names
    )

please give me advice.

**Comment by Yujin-Yujin** [comment]
Additionally, I tried

    for (n,p) in model.named_parameters():
          print(n, p.requires_grad)

then, I got the log that requires_grad of all the layers bias, weight is True. Is that correct?
Shouldn't be False for the all the layers?

However, even I tried 

   fusion_path = ="/home/test/fusion/siqa_defaut,a,b,c,d"
   model.load_adapter_fusion(fusion_path)
   model.freeze_model(freeze=True)

It didn't go better than before, still not loading the fusion.

**Comment by Yujin-Yujin** [comment]
Now I've tried code below, and it worked!
I removed **,config=PfeifferConfig(), with_head=False** from the model.load_adpater input.
I wonder this is correct because when I train the fusion I loaded adapter with **,config=PfeifferConfig(), with_head=False**.
Then should I load adapter which I trained without **,config=PfeifferConfig(), with_head=False**?

    model.load_adapter("/home/test/siqa_default", "text_task")
    model.load_adapter("/home/test/a", "text_task")
    model.load_adapter("/home/test/b", "text_task")
    model.load_adapter("/home/test/c", "text_task")
    model.load_adapter("/home/test/d", "text_task")
    
    adapter_names = [
            [
                "siqa_default",
                "a",
                "b",
                "c",
                "d"
            ]
        ]

    # pre-trained fusion_path 
        fusion_path = ="/home/test/fusion/siqa_defaut,a,b,c,d"
        model.load_adapter_fusion(fusion_path)

    # test_dataset
    test_dataset = (
        MultipleChoiceDataset(
            data_dir=data_args.data_dir,
            tokenizer=tokenizer,
            task=task_type,
            max_seq_length=data_args.max_seq_length,
            overwrite_cache=data_args.overwrite_cache,
            mode=Split.test,
        )
    )

def compute_metrics(p: EvalPrediction) -> Dict:
    preds = np.argmax(p.predictions, axis=1)
    return {"acc": simple_accuracy(preds, p.label_ids)}

# Initialize our Trainer
trainer = Trainer(
    model=model,
    args=training_args,
    compute_metrics=compute_metrics,
    adapter_names=adapter_names
)

**Comment by calpt** [comment]
Thanks for the updates!

> then, I got the log that requires_grad of all the layers bias, weight is True. Is that correct?
Shouldn't be False for the all the layers?

After you called `train_fusion()` this should indeed be `False` for all weights except the fusion weights. I tried to reproduce what you describe using similar code but didn't see the same results. For me, the fusion layers are successfully loaded and `requires_grad` is correctly set to `False` for all other weights. This is the exact code I used:

```python
import logging

from transformers import BertForSequenceClassification, PfeifferConfig


logging.basicConfig(level=logging.INFO)

model = BertForSequenceClassification.from_pretrained("bert-base-uncased")

model.load_adapter("multinli@ukp", "text_task", config=PfeifferConfig(), with_head=False)
model.load_adapter("./zzzz/a", config=PfeifferConfig(), with_head=False)

adapter_names = [
    [
        "mnli",
        "a"
    ]
]

# both works
# model.add_fusion(adapter_names[0], "dynamic")
model.load_adapter_fusion("./zzzz")
model.train_fusion(adapter_names)

for (n, p) in model.named_parameters():
    print(n, p.requires_grad)
```

showing this in the end:

```
...
bert.encoder.layer.11.output.adapter_fusion_layer.mnli,a.query.weight True
bert.encoder.layer.11.output.adapter_fusion_layer.mnli,a.query.bias True
bert.encoder.layer.11.output.adapter_fusion_layer.mnli,a.key.weight True
bert.encoder.layer.11.output.adapter_fusion_layer.mnli,a.key.bias True
bert.encoder.layer.11.output.adapter_fusion_layer.mnli,a.value.weight True
bert.encoder.layer.11.output.layer_text_task_adapters.mnli.adapter_down.0.weight False
bert.encoder.layer.11.output.layer_text_task_adapters.mnli.adapter_down.0.bias False
bert.encoder.layer.11.output.layer_text_task_adapters.mnli.adapter_up.weight False
bert.encoder.layer.11.output.layer_text_task_adapters.mnli.adapter_up.bias False
bert.encoder.layer.11.output.layer_text_task_adapters.a.adapter_down.0.weight False
bert.encoder.layer.11.output.layer_text_task_adapters.a.adapter_down.0.bias False
bert.encoder.layer.11.output.layer_text_task_adapters.a.adapter_up.weight False
bert.encoder.layer.11.output.layer_text_task_adapters.a.adapter_up.bias False
...
```

> I removed ,config=PfeifferConfig(), with_head=False from the model.load_adpater input.
I wonder this is correct because when I train the fusion I loaded adapter with ,config=PfeifferConfig(), with_head=False.
Then should I load adapter which I trained without ,config=PfeifferConfig(), with_head=False?

This also sounds very weird. You can always remove `config=PfeifferConfig()` when loading an adapter from a local path, the correct config will be loaded automatically from the folder. But it shouldn't break things if you still specify it. Does it only work if you remove _both_ parameters or is one of them sufficient?
Maybe it's also relevant how your `model` instance is defined, you didn't show this up to now.
